File: LeetCode/1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.java
class Solution {
    public int[] finalPrices(int[] prices) {
        int n = prices.length;
        Stack<Integer> st = new Stack<>();
        for (int i = 0; i < n; i++) {
            while (!st.isEmpty() && prices[i] <= prices[st.peek()]) {
                prices[st.peek()] -= prices[i];
                st.pop();
            }
            st.push(i);
        }
        return prices;
    }
}




// class Solution {
//     public int[] finalPrices(int[] prices) {
//         int n = prices.length;
//         for (int i = 0; i < n; i++) {
//             for (int j = i + 1; j < n; j++) {
//                 if (prices[i] >= prices[j]) {
//                     prices[i] -= prices[j];
//                     break;
//                 }
//             }
//         }
//         return prices;
//     }
// }

File: LeetCode/1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        n = len(prices)
        st = []
        for i in range(n):
            while st and prices[i] <= prices[st[-1]]:
                prices[st[-1]] -= prices[i]
                st.pop()
            st.append(i)
        return prices




# class Solution:
#     def finalPrices(self, prices: List[int]) -> List[int]:
#         n = len(prices)
#         for i in range(n):
#             for j in range(i+1, n):
#                 if prices[i] >= prices[j]:
#                     prices[i] -= prices[j]
#                     break
#         return prices